examples/src/main/java/com/squareup/moshi/recipes/TypeFailSafeJsonAdapterFactory.java

```diff
@@ -0,0 +1,102 @@
+/*
+ * Copyright (C) 2017 Square, Inc.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *      http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package com.squareup.moshi.recipes;
+
+import com.squareup.moshi.JsonAdapter;
+import com.squareup.moshi.JsonDataException;
+import com.squareup.moshi.JsonQualifier;
+import com.squareup.moshi.JsonReader;
+import com.squareup.moshi.JsonWriter;
+import com.squareup.moshi.Moshi;
+import java.io.IOException;
+import java.lang.annotation.Annotation;
+import java.lang.annotation.Retention;
+import java.lang.reflect.Type;
+import java.util.Collections;
+import java.util.LinkedHashSet;
+import java.util.Locale;
+import java.util.Set;
+
+import static java.lang.annotation.RetentionPolicy.RUNTIME;
+
+final class TypeFailSafeJsonAdapterFactory implements JsonAdapter.Factory {
+  private static final class Train {
+    @TypeFailSafe long optional_id;
+  }
+
+  public static void main(String[] args) {
+    Moshi moshi = new Moshi.Builder().add(new TypeFailSafeJsonAdapterFactory()).build();
+    JsonAdapter<Train> adapter = moshi.adapter(Train.class);
+    Train out;
+    try {
+      out = adapter.fromJson("{\"optional_id\":\"eighty\"}");
+    } catch (IOException e) {
+      throw new AssertionError(e);
+    }
+    System.out.println(out.optional_id);
+  }
+
+  @Retention(RUNTIME) @JsonQualifier @interface TypeFailSafe {
+  }
+
+  @Override public JsonAdapter<?> create(final Type type, Set<? extends Annotation> annotations,
+      com.squareup.moshi.Moshi moshi) {
+    Set<? extends Annotation> delegateAnnotations =
+        delegateAnnotations(annotations, TypeFailSafe.class);
+    if (delegateAnnotations == null) {
+      return null;
+    }
+    final JsonAdapter<Object> delegate = moshi.nextAdapter(this, type, delegateAnnotations);
+    return new JsonAdapter<Object>() {
+      @Override public Object fromJson(JsonReader reader) throws IOException {
+        try {
+          return delegate.fromJson(reader);
+        } catch (JsonDataException e) {
+          reader.skipValue();
```

Comment:
Unsure how I feel about this. It won’t do the right thing if the matched type isn’t a primitive or string because the exception might be thrown by a deeper method.